HLTrigger/btau/src/HLTDisplacedtktktkVtxProducer.cc

```diff
@@ -0,0 +1,280 @@
+
+{
+  produces<VertexCollection>();
+}
+
+
+HLTDisplacedtktktkVtxProducer::~HLTDisplacedtktktkVtxProducer() = default;
+
+void
+HLTDisplacedtktktkVtxProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("Src",edm::InputTag("hltL3MuonCandidates"));
+  desc.add<edm::InputTag>("PreviousCandTag",edm::InputTag(""));
+  desc.add<double>("MaxEtaTk",2.5);
+  desc.add<double>("MinPtResTk1",0.0);
+  desc.add<double>("MinPtResTk2",0.0);
+  desc.add<double>("MinPtThirdTk",0.0);
+  desc.add<double>("MinPtRes",0.0);
+  desc.add<double>("MinPtTri",0.0);
+  desc.add<double>("MinInvMassRes",1.0);
+  desc.add<double>("MaxInvMassRes",20.0);
+  desc.add<double>("MinInvMass",1.0);
+  desc.add<double>("MaxInvMass",20.0);
+  desc.add<double>("massParticleRes1",0.4937);
+  desc.add<double>("massParticleRes2",0.4937);
+  desc.add<double>("massParticle3",0.1396);
+  desc.add<int>("ChargeOpt",-1);
+  desc.add<int>("ResOpt",1);
+  desc.add<int>("triggerTypeDaughters",0);
+
+  descriptions.add("hltDisplacedtktktkVtxProducer", desc);
+}
+
+// ------------ method called once each job just before starting event loop  ------------
+void HLTDisplacedtktktkVtxProducer::beginJob()
+{
+
+}
+
+// ------------ method called once each job just after ending the event loop  ------------
+void HLTDisplacedtktktkVtxProducer::endJob() 
+{
+ 	
+}
+
+// ------------ method called on each new Event  ------------
+void HLTDisplacedtktktkVtxProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  double firstTrackMass = massParticle1_;
+  double secondTrackMass = massParticle2_;
+  double thirdTrackMass = massParticle3_;
+  double firstTrackPt = minPtTk1_;
+  double secondTrackPt = minPtTk2_;
+  double thirdTrackPt = minPtTk3_;
+  if(resOpt_<=0 && massParticle1_!=massParticle2_)
+    {
+      if(massParticle1_==massParticle3_) 
+        {
+          std::swap(secondTrackMass, thirdTrackMass);
+          std::swap(secondTrackPt, thirdTrackPt);
+        }
+      if(massParticle2_==massParticle3_) 
+        {
+          std::swap(firstTrackMass, thirdTrackMass);
+          std::swap(firstTrackPt, thirdTrackPt);
+        }
+    }
+  double const firstTrackMass2 = firstTrackMass*firstTrackMass;
+  double const secondTrackMass2 = secondTrackMass*secondTrackMass;
+  double const thirdTrackMass2 = thirdTrackMass*thirdTrackMass;
+
+  // get hold of track trks
+  Handle<RecoChargedCandidateCollection> trackcands;
+  iEvent.getByToken(srcToken_,trackcands);
+	
+  //get the transient track builder:
+  edm::ESHandle<TransientTrackBuilder> theB;
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder",theB);
+
+  std::unique_ptr<VertexCollection> vertexCollection(new VertexCollection());
+
+  // look at all trackcands,  check cuts and make vertices
+  double e1,e2,e3;
+  Particle::LorentzVector p,p1,p2,p3,pres;
+
+  if ( trackcands->size()  < 3 )   return;
```

Comment:
move up to where you retrieve the track collection

Reply:
changed